samples/subsys/usb/hid-mouse/src/main.c

```diff
@@ -0,0 +1,139 @@
+#endif
+
+/* change this to use another GPIO pin */
+#ifdef SW0_GPIO_PIN
+#define PIN     SW0_GPIO_PIN
+#else
+#error SW0_GPIO_PIN needs to be set in board.h
+#endif
+
+/*
+ * Usage Page (Desktop),              ; Generic desktop controls (01h)
+ * Usage (Mouse),                     ; Mouse (02h, application collection)
+ * Collection (Application),
+ *    Usage (Pointer),                ; Pointer (01h, physical collection)
+ *    Collection (Physical),
+ *        Usage Page (Button),        ; Button (09h)
+ *        Usage Minimum (01h),
+ *        Usage Maximum (03h),
+ *        Logical Minimum (0),
+ *        Logical Maximum (1),
+ *        Report Count (3),
+ *        Report Size (1),
+ *        Input (Variable),
+ *        Report Count (1),
+ *        Report Size (5),
+ *        Input (Constant),
+ *        Usage Page (Desktop),       ; Generic desktop controls (01h)
+ *        Usage (X),                  ; X (30h, dynamic value)
+ *        Usage (Y),                  ; Y (31h, dynamic value)
+ *        Usage (Wheel),              ; Wheel (38h, dynamic value)
+ *        Logical Minimum (-127),
+ *        Logical Maximum (127),
+ *        Report Size (8),
+ *        Report Count (3),
+ *        Input (Variable, Relative),
+ *    End Collection,
+ * End Collection
+ */
+
+static const u8_t hid_report_desc[] = {
+	0x05, 0x01, 0x09, 0x02, 0xa1, 0x01, 0x09, 0x01,
+	0xa1, 0x00, 0x05, 0x09, 0x19, 0x01, 0x29, 0x03,
+	0x15, 0x00, 0x25, 0x01, 0x95, 0x03, 0x75, 0x01,
+	0x81, 0x02, 0x95, 0x01, 0x75, 0x05, 0x81, 0x01,
+	0x05, 0x01, 0x09, 0x30, 0x09, 0x31, 0x09, 0x38,
+	0x15, 0x81, 0x25, 0x7f, 0x75, 0x08, 0x95, 0x03,
+	0x81, 0x06, 0xc0, 0xc0,
+};
+
+static int get_report_cb(struct usb_setup_packet *setup, s32_t *len,
+		  u8_t **data)
+{
+	*len = sizeof(hid_report_desc);
+	*data = hid_report_desc;
+
+	return 0;
+}
+
+static struct hid_ops ops = {
+	.get_report = get_report_cb,
+};
+
+static int def_val;
+static volatile int status, change;
+
+#define MOUSE_BTN_LEFT			BIT(0)
+#define MOUSE_BTN_RIGHT			BIT(1)
+#define MOUSE_BTN_MIDDLE		BIT(2)
+
+void button_pressed(struct device *gpio, struct gpio_callback *cb,
+		    u32_t pins)
+{
+	int cur_val, state;
+
+	gpio_pin_read(gpio, PIN, &cur_val);
+	if (def_val != cur_val) {
+		state |= MOUSE_BTN_LEFT;
+	} else {
+		state &= ~MOUSE_BTN_LEFT;
+	}
+
+	if (state != status) {
+		status = state;
+		change = 1;
+	}
+}
+
+void main(void)
+{
+	struct device *gpio;
+	struct gpio_callback gpio_cb;
+	u8_t report[4] = { 0x00 };
+
+	gpio = device_get_binding(PORT);
+	if (!gpio) {
+		printk("error\n");
+		return;
+	}
+
+	gpio_pin_configure(gpio, PIN,
+			GPIO_DIR_IN | GPIO_INT |
+			GPIO_INT_EDGE | GPIO_INT_DOUBLE_EDGE);
+
+	gpio_pin_read(gpio, PIN, &def_val);
+	gpio_init_callback(&gpio_cb, button_pressed, BIT(PIN));
+
+	gpio_add_callback(gpio, &gpio_cb);
+	gpio_pin_enable_callback(gpio, PIN);
+
+	usb_hid_register_device(hid_report_desc, sizeof(hid_report_desc), &ops);
+	usb_hid_init();
+
+	while (true) {
+		if (change) {
```

Comment:
A semaphore would be nicer instead of a global variable.

Reply:
got it, i will fix later